### PR TITLE
Improve cache usage in CI

### LINKED ISSUE #7677: Improve cache usage in CI
### Is your feature request related to a problem or challenge?

Currently a GA workflow defined in `rust.yml` uses `actions/cache`.
But there are some problems.

1. Cache is not actually used. See [this warning message](https://github.com/apache/arrow-datafusion/actions/runs/6327689637/job/17184438059#step:14:3). This means, the cache is not saves because the path is wrong. 
2. All the caches have the same key. So once a job saves a cache, other jobs fail to save their caches. Restoring a cache by the key will succeed but the contents of the cache might not be what a job want.
3. Caches are not updated because the key is constant (`cargo-cache-`).

So I think we have a chance to improve cache usage in CI.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sarutak**
## Which issue does this PR close?

Closes #7677

## Rationale for this change
Making the CI faster accelerate development.

## What changes are included in this PR?
This PR includes improvement of cache usage in CI.
In the current workflow, the critical path seems `cargo check` to `verify benchmark results (amd64)`. So `actions/cache` is used only for the jobs.
Cached contents are also revised.

## Are these changes tested?
Done by CI.
With my experiment, the total duration of workflow defined by `rust.yml` was about 30 mins.
https://github.com/sarutak/arrow-datafusion/actions/runs/6329626199

## Are there any user-facing changes?
No.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -47,18 +47,24 @@ jobs:
...
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            ./target/
+            ./datafusion-cli/target/
+          # this key equals the ones on `linux-build-lib` for re-use
+          key: cargo-cache-benchmark-${{ hashFiles('datafusion/**/Cargo.toml', 'benchmarks/Cargo.toml', 'datafusion-cli/Cargo.toml') }}
```
Comment (alamb) [inline .github/workflows/rust.yml:66]: This cache key doesn't change unless we update the dependencies, right? What about the local code changes?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -241,9 +229,14 @@ jobs:
...
-          path: /github/home/.cargo
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            ./target/
           # this key equals the ones on `linux-build-lib` for re-use
-          key: cargo-cache-
+          key: cargo-cache-benchmark-${{ hashFiles('datafusion/**/Cargo.toml', 'benchmarks/Cargo.toml') }}
```
Comment (alamb) [inline .github/workflows/rust.yml:239]: The key above also includes `'datafusion-cli/Cargo.toml'` -- is in intentional that these keys do not match?

```
          key: cargo-cache-benchmark-${{ hashFiles('datafusion/**/Cargo.toml', 'benchmarks/Cargo.toml', 'datafusion-cli/Cargo.toml') }}
```

**Review by alamb — APPROVED** [review]
Nice work @sarutak  -- thank you very much 

I will say I don't really understand how the cache action works / doesn't work for Rust projects. However, I did an unscientific spot check on PR CI timings for  this change, and my conclusion is that these data show a significant CI timing improvement and thus we can / should merge this PR. 

This PR:
Screenshot 2023-09-28 at 2 09 09 PM

https://github.com/apache/arrow-datafusion/pull/7671
Screenshot 2023-09-28 at 2 09 38 PM


https://github.com/apache/arrow-datafusion/pull/7669

Screenshot 2023-09-28 at 2 10 46 PM


https://github.com/apache/arrow-datafusion/pull/7663

Screenshot 2023-09-28 at 2 11 41 PM

File: .github/workflows/rust.yml (hunk 1 above)
Comment (sarutak) [reply .github/workflows/rust.yml:66]: Sorry for the late reply.
> This cache key doesn't change unless we update the dependencies, right?

Yes.

> What about the local code changes?

Local code change (except for Cargo.toml specified in `hashFiles`) doesn't affect the cache.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (sarutak) [reply .github/workflows/rust.yml:239]: Yes, it's intended.
This job (verify benchmark results) seems not to touch `datafusion-cli`, so this cache should not be stale on changing `datafusion-cli/Cargo.toml`.

**Pull request merged**